File: packages/contracts/scripts/fixture-verdicts.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { validateContract, type ContractName } from "../src/validator";

const root = resolve(import.meta.dir, "../../..");
const names: ContractName[] = ["aiTask", "content", "event", "operation"];
const verdicts: Record<string, boolean> = {};

for (const name of names) {
  for (const kind of ["valid", "invalid"] as const) {
    const file = name === "aiTask" ? "ai-task" : name;
    const value = JSON.parse(
      readFileSync(resolve(root, `docs/design/quality/fixtures/${file}.${kind}.json`), "utf8"),
    );
    verdicts[`${file}.${kind}`] = validateContract(name, value).valid;
  }
}
process.stdout.write(`${JSON.stringify(verdicts)}\n`);

File: crates/contracts/src/bin/fixture_verdicts.rs
use std::{collections::BTreeMap, fs, path::PathBuf};

use adoc_contracts::{ContractName, validator};

fn main() -> anyhow::Result<()> {
    let root = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../..");
    let contracts = [
        ("ai-task", ContractName::AiTask),
        ("content", ContractName::Content),
        ("event", ContractName::Event),
        ("operation", ContractName::Operation),
    ];
    let mut verdicts = BTreeMap::new();
    for (file, contract) in contracts {
        let compiled = validator(contract)?;
        for kind in ["valid", "invalid"] {
            let path = root.join(format!("docs/design/quality/fixtures/{file}.{kind}.json"));
            let value = serde_json::from_str(&fs::read_to_string(path)?)?;
            verdicts.insert(format!("{file}.{kind}"), compiled.is_valid(&value));
        }
    }
    println!("{}", serde_json::to_string(&verdicts)?);
    Ok(())
}
